doc/development/guide/architecture.rst

```diff
@@ -0,0 +1,196 @@
+utilize remote quantum devices and simulators.
+
+Components of PennyLane
+#######################
+
+.. image:: architecture_diagram.png
+    :width: 800px
+
+Devices
+*******
+
+In PennyLane, the abstraction of a quantum computation device is encompassed
+within the :class:`~.Device` class, making it one of the basic components of
+the library. The :class:`~.Device` class provides a common API for accessing quantum
+devices, independent of both the type of device (both simulators and hardware are supported),
+as well as the quantum information model used.
+
+In particular, the :class:`~.Device` class provides a common API for:
+
+* Initializing various device parameters such as the number of samples/shots,
+
+* Executing quantum circuits,
+
+* Retrieving measurement and state results (including samples, measurement
+  statistics, and probabilities).
+
+There are also device subclasses available, containing shared logic for
+particular types of devices.  For example, qubit-based devices can inherit from
+the :class:`~.QubitDevice` class, easing development.
+
+To register a new device with PennyLane, they must register an `entry point
+<https://packaging.python.org/specifications/entry-points/>`__under the `pennylane.plugins`
+namespace using Setuptools. Once registered, the device can be instantiated using the :func:`~.device`
+loader function.
+
+A Python package that registers one or more PennyLane device is known as a *plugin*. For more details
+on plugins and devices, see :doc:`/development/plugins`.
+
+The purpose of the ``Device`` class can be summarized as:
+
+* Providing a common API to execute a quantum circuit and request
+  the measurement of the associated observable.
+* Providing an easy way of developing a new device for PennyLane
+
+QNodes
+******
+
+A  quantum node or QNode (represented by a subclass of
+:class:`~.BaseQNode`) is an encapsulation of a function
+:math:`f(x;\theta)=R^m\rightarrow R^n` that is executed using quantum
+information processing on a quantum device.
+
+Apart from incorporating quantum functions, QNodes also offer custom
+quantum differentiation rules. Using the so-called `parameter-shift rules
+<https://pennylane.ai/qml/glossary/parameter_shift.html>`__, many quantum
+functions can be expressed through the linear combination of other quantum
+functions. As these rules allow quantum gradients to be obtained from
+QNodes, hybrid computations may include QNodes as part of training deep
+learnings models.
+
+These QNode types are available to users through the :func:`~.qnode` decorator by
+passing the user-facing ``diff_method`` option. This decorator then uses the
+:func:`~.QNode` constructor function to create the specific type of qnode based on
+the device, interface, and quantum function. If ``diff_method`` option is not
+provided, the QNode constructor function attempts to determine the ``"best"``
+differentiation method, based on the available device and interface.
```

Comment:
Minor suggestion: might be good to first introduce the QNode function before mentioning `diff_method`?

```suggestion
Users don't typically instantiate QNodes directly---instead, the :func:`~.qnode` decorator or
:func:`~.QNode` constructor function automates the process of creating a QNode from a provided
quantum function and device. The constructor attempts to determine the ``"best"`` QNode
subclass/differentiation method for the provided device and interface. For more fine-grained control,
the differentiation method can be specified directly via the ``diff_method`` option.
```
(feel free to modify!)